File: backend/src/crypto/card.rs
use crate::crypto::cipher::{open, seal};
use crate::crypto::identity::Identity;
use crate::crypto::session::Session;
use crate::error::{PeersError, Result};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::collections::HashMap;

/// Domain separation string for the identity-card signature.
const CARD_DOMAIN: &[u8] = b"peers/v1/card";
/// Wire format version of the sealed envelope.
const ENVELOPE_VERSION: u8 = 0x01;
/// Cap on recipients per message (defends the header against bloat).
const MAX_RECIPIENTS: usize = 64;

/// Self-authenticating identity card: the sender's Ed25519 public key, the
/// X25519 key used for ECDH, and an Ed25519 signature binding them. The
/// signature is over SHA-256("peers/v1/card" || x25519_pub), so anyone can
/// verify the card and cache (peer id -> x25519 key) for later sends.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PeerCard {
    pub ed_pub: [u8; 32],
    pub x25519_pub: [u8; 32],
    pub sig: Vec<u8>,
}

fn sign_message(x25519_pub: &[u8; 32]) -> [u8; 32] {
    let mut h = Sha256::new();
    h.update(CARD_DOMAIN);
    h.update(x25519_pub);
    h.finalize().into()
}

impl PeerCard {
    pub fn sign(identity: &Identity) -> Result<Self> {
        let x25519_pub = identity.x25519_public();
        let msg = sign_message(&x25519_pub);
        let sig = identity
            .keypair
            .sign(&msg)
            .map_err(|e| PeersError::Crypto(format!("card sign: {e}")))?;
        Ok(Self {
            ed_pub: identity.ed25519_public()?,
            x25519_pub,
            sig,
        })
    }

    /// Verifies the card's signature. The X25519 key is trusted after this.
    pub fn verify(&self) -> Result<()> {
        let pk = libp2p::identity::ed25519::PublicKey::try_from_bytes(&self.ed_pub)
            .map_err(|e| PeersError::Crypto(format!("card pubkey: {e}")))?;
        if !pk.verify(&sign_message(&self.x25519_pub), &self.sig) {
            return Err(PeersError::Crypto("card signature invalid".into()));
        }
        Ok(())
    }
}

/// Per-contact E2E state: one session (hash-chain + replay window) per
/// recipient X25519 key, plus the identity cards we have validated so far.
/// Memory-only for now; persisted contacts come with M4.
#[derive(Clone)]
pub struct SessionDir {
    sessions: HashMap<[u8; 32], Session>,
    contacts: HashMap<String, [u8; 32]>,
}

impl Default for SessionDir {
    fn default() -> Self {
        Self::new()
    }
}

impl SessionDir {
    pub fn new() -> Self {
        Self {
            sessions: HashMap::new(),
            contacts: HashMap::new(),
        }
    }

    fn session_for(&mut self, identity: &Identity, their_pub: [u8; 32]) -> Result<&mut Session> {
        if let std::collections::hash_map::Entry::Vacant(e) = self.sessions.entry(their_pub) {
            e.insert(Session::new(&identity.x25519_secret, their_pub)?);
        }
        Ok(self.sessions.get_mut(&their_pub).expect("just inserted"))
    }

    /// Remembers a validated card so future sends can encrypt to this peer.
    pub fn remember_contact(&mut self, peer_id: &str, card: &PeerCard) {
        self.contacts.insert(peer_id.to_string(), card.x25519_pub);
    }

    /// Snapshots the dir (per-contact sessions + contact x25519 keys) so it
    /// can be sealed to disk and restored after a restart. The secrets ride
    /// inside the state envelope, which is password-encrypted at rest.
    pub fn export(
        &self,
    ) -> (
        Vec<(Vec<u8>, crate::crypto::session::SessionState)>,
        Vec<(String, Vec<u8>)>,
    ) {
        let sessions = self
            .sessions
            .iter()
            .map(|(k, s)| (k.to_vec(), s.export()))
            .collect();
        let contacts = self
            .contacts
            .iter()
            .map(|(k, v)| (k.clone(), v.to_vec()))
            .collect();
        (sessions, contacts)
    }

    /// Restores state written by [`SessionDir::export`] into an empty dir.
    /// Existing contacts/sessions are kept; restored ones take precedence.
    pub fn restore(
        &mut self,
        sessions: &[(Vec<u8>, crate::crypto::session::SessionState)],
        contacts: &[(String, Vec<u8>)],
    ) {
        for (key, state) in sessions {
            if let Ok(key) = <[u8; 32]>::try_from(key.as_slice()) {
                self.sessions.insert(key, Session::import(state));
            }
        }
        for (peer, key) in contacts {
            if let Ok(key) = <[u8; 32]>::try_from(key.as_slice()) {
                self.contacts.insert(peer.clone(), key);
            }
        }
    }

    /// All X25519 keys we can currently encrypt to.
    pub fn recipient_keys(&self) -> Vec<[u8; 32]> {
        self.contacts.values().copied().collect()
    }

    /// Seals `plaintext` for every listed recipient. Payload layout:
    ///
    /// ```text
    /// [ver 1][ed_pub 32][x25519 32][sig 64][n u16 BE]
    ///   n x [rcpt_x25519 32][ct_len u32 BE][seq 8][nonce 12][ct+tag]
    /// ```
    ///
    /// `aad` binds the ciphertext to its context (channel id).
    pub fn seal(
        &mut self,
        identity: &Identity,
        recipients: &[[u8; 32]],
        aad: &[u8],
        plaintext: &[u8],
    ) -> Result<Vec<u8>> {
        if recipients.is_empty() {
            return Err(PeersError::Other(
                "no recipients with validated identity cards yet".into(),
            ));
        }
        if recipients.len() > MAX_RECIPIENTS {
            return Err(PeersError::MessageTooLarge);
        }
        let card = PeerCard::sign(identity)?;

        let mut copies = Vec::with_capacity(recipients.len() * 68 + plaintext.len());
        for rcpt in recipients {
            let session = self.session_for(identity, *rcpt)?;
            let sealed = seal(session, aad, plaintext)?;
            copies.extend_from_slice(rcpt);
            copies.extend_from_slice(&(sealed.len() as u32).to_be_bytes());
            copies.extend_from_slice(&sealed);
        }

        let mut out = Vec::with_capacity(1 + 32 + 32 + 64 + 2 + copies.len());
        out.push(ENVELOPE_VERSION);
        out.extend_from_slice(&card.ed_pub);
        out.extend_from_slice(&card.x25519_pub);
        out.extend_from_slice(&card.sig);
        out.extend_from_slice(&(recipients.len() as u16).to_be_bytes());
        out.extend_from_slice(&copies);
        Ok(out)
    }

    /// Opens a `seal`-produced envelope addressed to us. Verifies the sender
    /// card, picks our copy, and decrypts with replay protection. On success
    /// the sender's card is trusted and cached, and the plaintext returned.
    pub fn open(&mut self, identity: &Identity, aad: &[u8], payload: &[u8]) -> Result<Vec<u8>> {
        let card = parse_card(payload)?;
        card.verify()?;

        if card.x25519_pub == identity.x25519_public() {
            return Err(PeersError::Other("message from ourselves".into()));
        }
        if payload.len() < 131 {
            return Err(PeersError::BadCipher);
        }

        let own = identity.x25519_public();
        let n = u16::from_be_bytes(
            payload[129..131]
                .try_into()
                .map_err(|_| PeersError::BadCipher)?,
        ) as usize;
        let mut rest = &payload[131..];
        for _ in 0..n {
            if rest.len() < 32 + 4 + 8 + 16 {
                return Err(PeersError::BadCipher);
            }
            let rcpt: [u8; 32] = rest[..32].try_into().map_err(|_| PeersError::BadCipher)?;
            let ct_len =
                u32::from_be_bytes(rest[32..36].try_into().map_err(|_| PeersError::BadCipher)?)
                    as usize;
            let copy_len = 32 + 4 + ct_len;
            if rest.len() < copy_len {
                return Err(PeersError::BadCipher);
            }
            let sealed = &rest[36..36 + ct_len];
            if rcpt == own {
                let session = self.session_for(identity, card.x25519_pub)?;
                let plaintext = open(session, aad, sealed)?;
                self.remember_contact(&card_contact_key(&card), &card);
                return Ok(plaintext);
            }
            rest = &rest[copy_len..];
        }
        Err(PeersError::NotAddressed)
    }
}

/// Contact map key for a card: the Ed25519-based peer id string.
fn card_contact_key(card: &PeerCard) -> String {
    // Stable pseudo-peer-id: "k" + base32 of the ed25519 key. The real peer
    // id (multihash) is not derivable from the raw key without extra deps.
    base32::encode(base32::Alphabet::Rfc4648 { padding: false }, &card.ed_pub)
}

fn parse_card(payload: &[u8]) -> Result<PeerCard> {
    let min = 1 + 32 + 32 + 64 + 2;
    if payload.len() < min {
        return Err(PeersError::BadCipher);
    }
    if payload[0] != ENVELOPE_VERSION {
        return Err(PeersError::BadCipher);
    }
    let ed_pub: [u8; 32] = payload[1..33]
        .try_into()
        .map_err(|_| PeersError::BadCipher)?;
    let x25519_pub: [u8; 32] = payload[33..65]
        .try_into()
        .map_err(|_| PeersError::BadCipher)?;
    let sig: Vec<u8> = payload[65..129].to_vec();
    Ok(PeerCard {
        ed_pub,
        x25519_pub,
        sig,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::crypto::identity::Identity;

    fn pair() -> (Identity, Identity) {
        (Identity::new().unwrap(), Identity::new().unwrap())
    }

    #[test]
    fn card_verifies() {
        let (a, _b) = pair();
        let card = PeerCard::sign(&a).unwrap();
        card.verify().unwrap();
    }

    #[test]
    fn tampered_card_rejected() {
        let (a, _b) = pair();
        let mut card = PeerCard::sign(&a).unwrap();
        card.sig[0] ^= 0xff;
        assert!(card.verify().is_err());
    }

    #[test]
    fn a_to_b_round_trip() {
        let (a, b) = pair();
        let b_pub = b.x25519_public();
        let mut da = SessionDir::new();
        let mut db = SessionDir::new();
        let payload = da
            .seal(&a, &[b_pub], b"channel/general", b"hello b")
            .unwrap();
        let got = db.open(&b, b"channel/general", &payload).unwrap();
        assert_eq!(got, b"hello b");
        // b can now reply: a's key is cached under b's contacts.
        let a_pub = a.x25519_public();
        assert!(db.recipient_keys().contains(&a_pub));
    }

    #[test]
    fn third_party_cannot_open() {
        let (a, b) = pair();
        let c = Identity::new().unwrap();
        let b_pub = b.x25519_public();
        let mut da = SessionDir::new();
        let mut dc = SessionDir::new();
        let payload = da.seal(&a, &[b_pub], b"ch", b"secret").unwrap();
        assert!(matches!(
            dc.open(&c, b"ch", &payload),
            Err(PeersError::NotAddressed) | Err(PeersError::Crypto(_))
        ));
    }

    #[test]
    fn wrong_aad_rejected() {
        let (a, b) = pair();
        let b_pub = b.x25519_public();
        let mut da = SessionDir::new();
        let mut db = SessionDir::new();
        let payload = da.seal(&a, &[b_pub], b"channel/one", b"secret").unwrap();
        assert!(db.open(&b, b"channel/other", &payload).is_err());
    }

    #[test]
    fn replay_rejected_across_copies() {
        let (a, b) = pair();
        let b_pub = b.x25519_public();
        let mut da = SessionDir::new();
        let mut db = SessionDir::new();
        let payload = da.seal(&a, &[b_pub], b"ch", b"first").unwrap();
        db.open(&b, b"ch", &payload).unwrap();
        // Same envelope replayed (even after a new message).
        let second = da.seal(&a, &[b_pub], b"ch", b"second").unwrap();
        db.open(&b, b"ch", &second).unwrap();
        assert!(matches!(
            db.open(&b, b"ch", &payload),
            Err(PeersError::Replay)
        ));
    }

    #[test]
    fn multi_recipient() {
        let (a, b) = pair();
        let c = Identity::new().unwrap();
        let mut da = SessionDir::new();
        let mut db = SessionDir::new();
        let mut dc = SessionDir::new();
        let payload = da
            .seal(&a, &[b.x25519_public(), c.x25519_public()], b"ch", b"both")
            .unwrap();
        assert_eq!(db.open(&b, b"ch", &payload).unwrap(), b"both");
        assert_eq!(dc.open(&c, b"ch", &payload).unwrap(), b"both");
    }

    #[test]
    fn tampered_envelope_rejected() {
        let (a, b) = pair();
        let b_pub = b.x25519_public();
        let mut da = SessionDir::new();
        let mut db = SessionDir::new();
        let mut payload = da.seal(&a, &[b_pub], b"ch", b"don't touch").unwrap();
        let last = payload.len() - 1;
        payload[last] ^= 0xff;
        assert!(db.open(&b, b"ch", &payload).is_err());
    }
}

File: packages/core/src/card.ts
import {ed25519} from '@noble/curves/ed25519.js';
import {sha256} from '@noble/hashes/sha2.js';
import {peersErr} from './error.js';
import type {Identity} from './identity.js';
import {derivePeerId} from './peerid.js';
import {Session, type SessionState} from './session.js';
import {open as openEnvelope, seal as sealEnvelope} from './cipher.js';
import {b64decode, b64encode, canonicalJson, concatBytes, utf8} from './util.js';

/** Domain separation string for the identity-card signature. */
const CARD_DOMAIN = 'peers/v1/card';
/** Wire format version of the sealed envelope. */
const ENVELOPE_VERSION = 0x01;
/** Cap on recipients per message (defends the header against bloat). */
const MAX_RECIPIENTS = 64;

/**
 * Self-authenticating identity card: the sender's Ed25519 public key, the
 * X25519 key used for ECDH, and an Ed25519 signature binding them. The
 * signature is over SHA-256("peers/v1/card" || x25519_pub), so anyone can
 * verify the card and cache (peer id -> x25519 key) for later sends.
 */
export interface PeerCard {
  edPub: Uint8Array;
  x25519Pub: Uint8Array;
  sig: Uint8Array;
}

function cardSignBytes(x25519Pub: Uint8Array): Uint8Array {
  return sha256(concatBytes(utf8(CARD_DOMAIN), x25519Pub));
}

export const PeerCard = {
  sign(identity: Identity): PeerCard {
    const x25519Pub = identity.xPublic();
    return {
      edPub: identity.edPublic(),
      x25519Pub,
      sig: identity.sign(cardSignBytes(x25519Pub)),
    };
  },

  /** Verifies the card's signature. The X25519 key is trusted after this. */
  verify(card: PeerCard): boolean {
    if (card.edPub.length !== 32 || card.x25519Pub.length !== 32) return false;
    try {
      return ed25519.verify(card.sig, cardSignBytes(card.x25519Pub), card.edPub);
    } catch {
      return false;
    }
  },

  /** The peer id implied by a verified card's embedded key. */
  peerIdOf(card: PeerCard): string {
    return derivePeerId(card.edPub);
  },
};

/**
 * Stable, playful auto-generated display name ("JuicyPear") seeded from
 * the peer id bytes, so every identity has a fun default without needing
 * to persist it.
 */
export function defaultDisplayName(peerIdBytes: Uint8Array): string {
  const ADJS = ['Juicy', 'Cosmic', 'Turbo', 'Silky', 'Crispy', 'Golden', 'Mellow', 'Fuzzy', 'Swift', 'Velvet', 'Spicy', 'Breezy'];
  const NOUNS = ['Pear', 'Comet', 'Duck', 'Ghost', 'Taco', 'Beetle', 'Cloud', 'Fox', 'Cactus', 'Orbit', 'Llama', 'Waffle'];
  const a = peerIdBytes[peerIdBytes.length - 1]! % ADJS.length;
  const n = peerIdBytes[0]! % NOUNS.length;
  return ADJS[a]! + NOUNS[n]!;
}

/**
 * Self-authenticating display profile: display name, "about" line and an
 * optional avatar blob hash, all signed by the identity so they can't be
 * impersonated. Rides alongside the peer card wherever contacts are
 * exchanged (member lists, join notices, profile notices).
 *
 * `sig` covers every other field via canonical JSON with this exact key
 * order and `sig: ""`.
 */
export interface SignedProfile {
  version: number;
  peerId: string;
  pubkey: number[];
  displayName: string;
  about: string;
  avatarHash: string | null;
  sig: string;
}

export const SignedProfile = {
  sign(identity: Identity, displayName: string, about: string, avatarHash: string | null): SignedProfile {
    const trimmed = displayName.trim();
    const p: SignedProfile = {
      version: 1,
      peerId: identity.peerId,
      pubkey: Array.from(identity.edPublic()),
      displayName: trimmed.length > 0 ? trimmed : defaultDisplayName(identity.peerIdBytes),
      about: about.trim(),
      avatarHash,
      sig: '',
    };
    p.sig = b64encode(identity.sign(canonicalJson(profileSignView(p))));
    return p;
  },

  /** Verifies the signature and that it is bound to `peerId`. */
  verify(p: SignedProfile): void {
    let ok = false;
    try {
      ok =
        p.pubkey.length === 32 &&
        ed25519.verify(b64decodeSig(p.sig), canonicalJson(profileSignView(p)), Uint8Array.from(p.pubkey));
    } catch {
      ok = false;
    }
    if (!ok) throw peersErr('Crypto', 'profile signature invalid');
    const derived = derivePeerId(Uint8Array.from(p.pubkey));
    if (derived !== p.peerId) throw peersErr('Crypto', 'profile peer id mismatch');
  },
};

/** The object exactly as it was serialized for signing (`sig: ""`). */
function profileSignView(p: SignedProfile): Record<string, unknown> {
  return {
    version: p.version,
    peerId: p.peerId,
    pubkey: p.pubkey,
    displayName: p.displayName,
    about: p.about,
    avatarHash: p.avatarHash,
    sig: '',
  };
}

function b64decodeSig(s: string): Uint8Array {
  return b64decode(s);
}

/**
 * Per-contact E2E state: one session per recipient X25519 key, plus the
 * identity cards validated so far. Sealed envelopes carry one copy of the
 * plaintext per recipient:
 *
 * ```text
 * [ver 1][ed_pub 32][x25519 32][sig 64][n u16 BE]
 *   n × [rcpt_x25519 32][ct_len u32 BE][seq 8][nonce-derived ct+tag]
 * ```
 *
 * `aad` binds each copy to its context (e.g. channel).
 */
export class SessionDir {
  private readonly sessions = new Map<string, Session>();
  private readonly contacts = new Map<string, Uint8Array>();

  constructor(private readonly identity: Identity) {}

  static restore(
    state: {sessions: [Uint8Array, SessionState][]; contacts: [string, Uint8Array][]},
    identity: Identity,
  ): SessionDir {
    const dir = new SessionDir(identity);
    for (const [key, s] of state.sessions) dir.sessions.set(toKey(key), Session.import(s));
    for (const [peer, key] of state.contacts) dir.contacts.set(peer, key.slice());
    return dir;
  }

  export(): {sessions: [Uint8Array, SessionState][]; contacts: [string, Uint8Array][]} {
    return {
      sessions: [...this.sessions].map(([k, s]) => [fromKey(k), s.export()]),
      contacts: [...this.contacts].map(([p, k]) => [p, k.slice()]),
    };
  }

  /** Remembers a validated card so future sends can encrypt to this peer. */
  rememberContact(peerId: string, card: PeerCard): void {
    this.contacts.set(peerId, card.x25519Pub.slice());
  }

  /** The X25519 key we can encrypt to for a peer id, if we hold their
   * validated card. `undefined` means keys were never exchanged. */
  recipientKey(peerId: string): Uint8Array | undefined {
    return this.contacts.get(peerId);
  }

  recipientKeys(): Uint8Array[] {
    return [...this.contacts.values()].map((k) => k.slice());
  }

  seal(recipients: Uint8Array[], aad: Uint8Array, plaintext: Uint8Array): Uint8Array {
    if (recipients.length === 0) {
      throw peersErr('Other', 'no recipients with validated identity cards yet');
    }
    if (recipients.length > MAX_RECIPIENTS) throw peersErr('MessageTooLarge');

    const card = PeerCard.sign(this.identity);
    const copies: Uint8Array[] = [];
    for (const rcpt of recipients) {
      const session = this.sessionFor(rcpt);
      const sealed = sealEnvelope(session, aad, plaintext);
      const head = new Uint8Array(36);
      head.set(rcpt, 0);
      new DataView(head.buffer).setUint32(32, sealed.length);
      copies.push(concatBytes(head, sealed));
    }

    const copiesLen = copies.reduce((n, c) => n + c.length, 0);
    const out = new Uint8Array(1 + 32 + 32 + 64 + 2 + copiesLen);
    const view = new DataView(out.buffer);
    let off = 0;
    out[0] = ENVELOPE_VERSION;
    off += 1;
    out.set(card.edPub, off); off += 32;
    out.set(card.x25519Pub, off); off += 32;
    out.set(card.sig, off); off += 64;
    view.setUint16(off, recipients.length); off += 2;
    for (const c of copies) {
      out.set(c, off);
      off += c.length;
    }
    return out;
  }

  /**
   * Opens an envelope addressed to us. Verifies the sender card, picks our
   * copy, decrypts with replay protection. On success the sender's card is
   * cached under their Ed25519-derived peer id.
   */
  open(aad: Uint8Array, payload: Uint8Array): Uint8Array {
    const card = parseCard(payload);
    if (!PeerCard.verify(card)) throw peersErr('BadCipher', 'card signature invalid');
    if (bytesEqual(card.x25519Pub, this.identity.xPublic())) {
      throw peersErr('Other', 'message from ourselves');
    }

    const view = new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
    const n = view.getUint16(129);
    let rest = payload.subarray(131);
    const own = this.identity.xPublic();
    for (let i = 0; i < n; i++) {
      if (rest.length < 32 + 4 + 8 + 16) throw peersErr('BadCipher');
      const rcpt = rest.slice(0, 32);
      const ctLen = new DataView(rest.buffer, rest.byteOffset, rest.byteLength).getUint32(32);
      const copyLen = 32 + 4 + ctLen;
      if (rest.length < copyLen) throw peersErr('BadCipher');
      const sealed = rest.subarray(36, 36 + ctLen);
      if (bytesEqual(rcpt, own)) {
        const session = this.sessionFor(card.x25519Pub);
        const plaintext = openEnvelope(session, aad, sealed);
        this.rememberContact(PeerCard.peerIdOf(card), card);
        return plaintext;
      }
      rest = rest.subarray(copyLen);
    }
    throw peersErr('NotAddressed');
  }

  private sessionFor(theirPub: Uint8Array): Session {
    const k = toKey(theirPub);
    let session = this.sessions.get(k);
    if (!session) {
      session = Session.agree(this.identity.xSecret, theirPub);
      this.sessions.set(k, session);
    }
    return session;
  }
}

/** Parses (does not verify) the sender's card from a sealed envelope. */
export function cardFromEnvelope(payload: Uint8Array): PeerCard {
  return parseCard(payload);
}

function parseCard(payload: Uint8Array): PeerCard {
  const min = 1 + 32 + 32 + 64 + 2;
  if (payload.length < min || payload[0] !== ENVELOPE_VERSION) throw peersErr('BadCipher');
  return {
    edPub: payload.slice(1, 33),
    x25519Pub: payload.slice(33, 65),
    sig: payload.slice(65, 129),
  };
}

function toKey(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString('base64');
}

function fromKey(s: string): Uint8Array {
  return Uint8Array.from(Buffer.from(s, 'base64'));
}

function bytesEqual(a: Uint8Array, b: Uint8Array): boolean {
  return a.length === b.length && a.every((v, i) => v === b[i]);
}
